carma_ros2_utils/src/lifecycle_component_wrapper.cpp

```diff
@@ -0,0 +1,410 @@
+
+  auto classes = loader->getAvailableClasses<rclcpp_components::NodeFactory>();
+  for (const auto & clazz : classes) {
+    RCLCPP_INFO(get_logger(), "Found class: %s", clazz.c_str());
+    if (clazz == class_name || clazz == fq_class_name) {
+      RCLCPP_INFO(get_logger(), "Instantiate class: %s", clazz.c_str());
+      return loader->createInstance<rclcpp_components::NodeFactory>(clazz);
+    }
+  }
+  return {};
+}
+
+rclcpp::NodeOptions
+LifecycleComponentWrapper::create_node_options(const std::shared_ptr<LoadNode::Request> request)
+{
+  std::vector<rclcpp::Parameter> parameters;
+  for (const auto & p : request->parameters) {
+    parameters.push_back(rclcpp::Parameter::from_parameter_msg(p));
+  }
+
+  std::vector<std::string> remap_rules;
+  remap_rules.reserve(request->remap_rules.size() * 2 + 1);
+  remap_rules.push_back("--ros-args");
+  for (const std::string & rule : request->remap_rules) {
+    remap_rules.push_back("-r");
+    remap_rules.push_back(rule);
+  }
+
+  if (!request->node_name.empty()) {
+    remap_rules.push_back("-r");
+    remap_rules.push_back("__node:=" + request->node_name);
+  }
+
+  if (!request->node_namespace.empty()) {
+    remap_rules.push_back("-r");
+    remap_rules.push_back("__ns:=" + request->node_namespace);
+  }
+
+  auto options = rclcpp::NodeOptions()
+    .use_global_arguments(false)
+    .parameter_overrides(parameters)
+    .arguments(remap_rules);
+
+  for (const auto & a : request->extra_arguments) {
+    const rclcpp::Parameter extra_argument = rclcpp::Parameter::from_parameter_msg(a);
+    if (extra_argument.get_name() == "use_intra_process_comms") {
+      if (extra_argument.get_type() != rclcpp::ParameterType::PARAMETER_BOOL) {
+        throw rclcpp_components::ComponentManagerException(
+                "Extra component argument 'use_intra_process_comms' must be a boolean");
+      }
+      options.use_intra_process_comms(extra_argument.get_value<bool>());
+    }
+  }
+
+  return options;
+}
+
+void
+LifecycleComponentWrapper::on_load_node(
+  const std::shared_ptr<rmw_request_id_t> request_header,
+  const std::shared_ptr<LoadNode::Request> request,
+  std::shared_ptr<LoadNode::Response> response, boost::optional<uint64_t> internal_id)
+{
+  (void) request_header;
+
+  ///// CARMA CHANGE /////
+  // If this was an external request and we are NOT in the ACTIVE state
+  // then we should not load the node and simply cache it for load on activate
+  if (!internal_id && get_current_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {
+    
+    RCLCPP_INFO_STREAM(get_logger(), "Got request to load node: " << request->node_name <<
+      " but we are not in the active state, caching for later lifecycle based activation.");
+
+    auto node_id = unique_id_++; // Store and update the unique id
+     
+    load_node_requests_.emplace(node_id, std::make_pair(*request_header, *request));
+
+    // NOTE: This is a bit of a hack, but we need to return a valid response
```

Comment:
Is there any sort of issue with this where a calling component might think the node is actually loaded and act on that assumption some how?

Reply:
Its certainly possible, but I think it must be left to the user to make that call. This small "hack" is required to allow the existing ros launch functions to interface with this component properly.

Reply:
I haven't had any issues with the existing ROS2 tool in ros launch or ros2 lifecycle.